go-controller/pkg/util/util.go

```diff
@@ -888,6 +888,19 @@ func GetEndpointsForService(endpointSlices []*discoveryv1.EndpointSlice, service
 	return globalEndpoints, localEndpoints, errors.Join(validationErrors...)
 }
 
+// FindServicePortForEndpointSlicePort returns the ServicePort that corresponds to an EndpointSlice port
+// by matching the port name and protocol. This is the canonical way to map EndpointSlice ports to
+// Service ports, as Kubernetes guarantees that ServicePort.Name matches EndpointPort.Name.
+func FindServicePortForEndpointSlicePort(service *corev1.Service, endpointslicePortName string, endpointslicePortProtocol corev1.Protocol) (corev1.ServicePort, error) {
+	for _, servicePort := range service.Spec.Ports {
+		if servicePort.Name == endpointslicePortName && servicePort.Protocol == endpointslicePortProtocol {
+			return servicePort, nil
+		}
+	}
+	return corev1.ServicePort{}, fmt.Errorf("service %s/%s has no port with name %s and protocol %s",
```

Comment:
NIT, in case the port name is an empty string:
```suggestion
	return corev1.ServicePort{}, fmt.Errorf("service %s/%s has no port with name %q and protocol %s",
```